Add api_info and api_modify modules

felixfontein opened the pull request:
##### SUMMARY
This adds two new modules, `api_info` and `api_modify`, which support querying and modifying certain paths.

The modules provide a high-level interface and only support some paths. The API information included is based on what I gathered from RouterOS 6.49. I'd like to have a better way to gather this data, but haven't been able to look into that yet.

The `api_modify ` module should be able to at least solve some of the problems mentioned in #33.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
api_info
api_modify

felixfontein commented:
Example:
```.yaml
- community.routeros.api_modify:
    path: ip dns static
    data:
      - address: 192.168.88.1
        name: router
      - address: 192.168.88.2
        name: server1
      - address: 192.168.88.3
        name: server2
    handle_absent_entries: remove
    handle_entries_content: remove
    ensure_order: true
```
This will ensure that `/ip dns static` has extractly three entries in the specified order of exactly the specified format.

NikolayDachev commented:
I'm not sure why  ```choices:``` list  must be in docs ? Is there anything specifics for those paths ? what I miss ?

NikolayDachev commented:
As general I absolutely like it :D 

More questions 

This NAT to WAN example .. make no sense to me, Also I get error for ``` "msg": "Unknown key \"name\" at index 1."``` which is ok since I'm not sure if ip firewall nat .. have attribute name ?! 

The example from this PR make more sense  (about DNS static entries )

From code prospective I think is ok

felixfontein commented:
It should have been `comment`, not `name` in that example... I've updated it and added another example.

`choices:` is kind of needed since the modules only support some pathes, namely the ones they "know how they work" and for which they have information in fields that exist, default values, etc.

NikolayDachev reviewed (approved):
Look ok to be merged in main

NikolayDachev commented:
note: Probably rebase will be needed today I merge "extended-query"

felixfontein commented:
Things this PR is still missing:
- [ ] A way to get hold of the API information. Right now it's manually compiling everything, but that's not really efficient and will likely cause problems already with RouterOS 7. Also there are quite some paths that aren't fully implemented yet due to me not knowing how to correctly support them, how exactly they work, etc.
- [ ] Having better data, like some options only work when other options have certain values, etc. Might also be related to the above point.
- [ ] Ordering for paths with stratified keys seems to have problems (maybe also for paths with primary keys?). For some firewall rules I had to run the module several times until they finally had the correct order. I'll need to debug this more to figure out what causes this and how to fix it.
- [ ] Not related directly to this PR, but a common problem for all `api*` modules: using non-ASCII characters. For example I had an `ü` in a comment for my home router, and the module crashed when trying to send that comment to RouterOS since it didn't like that letter. I was able to set it in the web UI though.

NikolayDachev commented:
For firewall order (not directly related) from ros prospective 

Let say we have 
```
 0    chain=forward action=accept log=no log-prefix="" 
 1    chain=forward action=accept log=no log-prefix="" 
 2    chain=forward action=accept log=no log-prefix="" 
```
If you want to change rule number 1 to action=drop **set** can be used

```
 /ip/firewall/filter/set action=drop numbers=1
```
result

```
 0    chain=forward action=accept log=no log-prefix="" 
 1    chain=forward action=drop log=no log-prefix="" 
 2    chain=forward action=accept log=no log-prefix="" 
```

@felixfontein  Not sure if this example can help, probably if you provide me more info ... ?

NikolayDachev commented:
For " Having better data, like some options only work when other options have certain values, etc. Might also be related to the above point."

I guess you know some routeros paths haves sub path, so is normal some attributes to depend on other (specially with sub paths) .. if I understand correctly the problem ?

felixfontein commented:
No, I meant more that for some values in one path some attributes only make sense in some case, like in `/ip service`, `certificate` and `tls-version` make sense for `www-ssl` and `api-ssl`, but not for `www`, `api`, `ftp`, ... There are more cases like this in other paths.

NikolayDachev commented:
"Also there are quite some paths that aren't fully implemented yet due to me not knowing how to correctly support them, how exactly they work, etc."

Let's clear the problems her first on ROS level then we can see what we can do with API, let me know if you have a issues with something specific, probably I can help 

Yes ROS7 don't have long term support version yet, we cannot predict every change so simple "disclaimer" in doc's will do the job for now , for the rest we can relay on module users feedback !

NikolayDachev commented:
> No, I meant more that for some values in one path some attributes only make sense in some case, like in `/ip service`, `certificate` and `tls-version` make sense for `www-ssl` and `api-ssl`, but not for `www`, `api`, `ftp`, ... There are more cases like this in other paths.

I think I understand the problem, unfortunately no clear solution,
1. "hardcoded check per case"
2. somehow routeros trap errors to be tracked .. probably 
3. no idea really

NikolayDachev commented:
@felixfontein  let me know if you need a help with something here

felixfontein commented:
> @felixfontein let me know if you need a help with something here

I was looking at how to automatically export information on the API structure from RouterOS recently (like which fields are supported for which path etc.), but I couldn't really find anything.

Some information seems to be available from the HTTP interface, like https://192.168.88.1/webfig/roteros.jg (base system) or https://192.168.88.1/webfig/wlan6.jg (there seems to be one such file per package that can be installed). This is unfortunately not JSON, but something a JavaScript interpreter understands, and I suspect it contains *some* information needed to render the web UI. *But* it doesn't seem to be everything, it probably works together with other internal data.

That's probably the most important thing left...

NikolayDachev commented:
Ros7 have REST api support, if you need something specific probably can look it with postman, I guess rest and routeros api are similar .

Not sure if api structure can be dump in both cases

felixfontein commented:
@NikolayDachev I haven't been able to really work on this, but I also don't see how to actually solve the remaining problems. What do you think of simply declaring this as "finished" for now and merging it and releasing it, so folks can start trying it out and maybe even extending/improving/... it for the use cases they have?

felixfontein commented:
I'll update the docs and mark this as ready once I've pushed that. Might not happen today, but very soon :)

felixfontein commented:
@NikolayDachev what do you think of the formulation I pushed?

NikolayDachev commented:
I will try to find a time during the weekend

NikolayDachev commented:
Looking good, just check my suggestion in code review .. I don't think is something really special.
Depend on your decision .. overall we can merge if no future changes are needed.

felixfontein commented:
@NikolayDachev in case you had a suggestion, you didn't add it to the review :)

NikolayDachev commented:
?

[NikolayDachev](https://github.com/NikolayDachev) started a review
---
Probably 'allows to retrieve information on specific paths that can' can be changed to somethin like

'...allows to retrieve information on specific predefined paths that can...' since we do that with _api_data (If I not mistake)
Overall we can keep it in this format .. this is just a suggestion !

I know this will make it little more complicated but if we have a option we can do auto generate list in docs section (I'm not sure if is possible to use [list] ?!).
However in all cases this will be a good feature, not something mandatory

NikolayDachev commented:
> @NikolayDachev in case you had a suggestion, you didn't add it to the review :)

sorry was in pending state

felixfontein commented on docs/docsite/rst/api-guide.rst:
```diff
@@ -8,7 +8,7 @@
 ========================================================
 
-You can use the :ref:`community.routeros.api module <ansible_collections.community.routeros.api_module>` to connect to a RouterOS device with the RouterOS API. A more specific module to modify certain entries is the :ref:`community.routeros.api_find_and_modify module <ansible_collections.community.routeros.api_find_and_modify_module>`. The :ref:`community.routeros.api_facts module <ansible_collec ...
+You can use the :ref:`community.routeros.api module <ansible_collections.community.routeros.api_module>` to connect to a RouterOS device with the RouterOS API. More specific module to modify certain entries are the :ref:`community.routeros.api_modify <ansible_collections.community.routeros.api_modify_module>` and :ref:`community.routeros.api_find_and_modify <ansible_collections.community.routeros ...
```
I updated the formulation in 6c1b859ae79d31e2d1a508febb4eb150ca843461. I'll try something for generating these lists, will push that later.

felixfontein commented:
Ok, ready for another round of review :)

NikolayDachev reviewed (approved):
look good to go

felixfontein commented:
Then let's do this! Thanks a lot for reviewing and testing this, @NikolayDachev!
I'll create a new release probably tomorrow, so folks can more easily try this out.
